frontends/concrete-python/concrete/fhe/mlir/processors/assign_bit_widths.py

```diff
@@ -13,266 +13,346 @@
 
 class AssignBitWidths(GraphProcessor):
     """
-    Assign a precision to all nodes inputs/output.
-
-    The precisions are compatible graph constraints and MLIR.
-    There are two modes:
-        - single precision: where all encrypted values have the same precision.
-        - multi precision: where encrypted values can have different precisions.
+    AssignBitWidths graph processor, to assign proper bit-widths to be compatible with FHE.
     """
 
     def __init__(self, single_precision=False):
         self.single_precision = single_precision
 
     def apply(self, graph: Graph):
-        nodes = graph.query_nodes()
-        for node in nodes:
-            assert isinstance(node.output.dtype, Integer)
-            node.properties["original_bit_width"] = node.output.dtype.bit_width
+        solver = z3.Solver()
 
-        if self.single_precision:
-            assign_single_precision(nodes)
-        else:
-            assign_multi_precision(graph, nodes)
+        bit_widths: Dict[Node, z3.Int] = {}
 
+        additional_constraints = AdditionalConstraints(solver, graph, bit_widths)
 
-def assign_single_precision(nodes: list[Node]):
-    """Assign one single encryption precision to all nodes."""
-    p = required_encrypted_bitwidth(nodes)
-    for node in nodes:
-        assign_precisions_1_node(node, p, p)
+        nodes = graph.query_nodes(ordered=True)
+        for i, node in enumerate(nodes):
+            bit_width = z3.Int(f"%{i}")
+            bit_widths[node] = bit_width
 
+            assert isinstance(node.output.dtype, Integer)
+            solver.add(bit_width >= node.output.dtype.bit_width)
 
-def assign_precisions_1_node(node: Node, output_p: int, inputs_p: int):
-    """Assign input/output precision to a single node.
+            additional_constraints.generate_for(node, bit_width)
 
-    Precision are adjusted to match different use, e.g. encrypted and constant case.
-    """
-    assert isinstance(node.output.dtype, Integer)
-    if node.output.is_encrypted:
-        node.output.dtype.bit_width = output_p
-    else:
-        node.output.dtype.bit_width = output_p + 1
+        if self.single_precision:
+            for i in range(len(nodes) - 1):
+                solver.add(bit_widths[nodes[i]] == bit_widths[nodes[i + 1]])
 
-    for value in node.inputs:
-        assert isinstance(value.dtype, Integer)
-        if value.is_encrypted:
-            value.dtype.bit_width = inputs_p
-        else:
-            value.dtype.bit_width = inputs_p + 1
+        optimizer = z3.Optimize()
+        optimizer.minimize(sum(bit_widths.values()))
```

Comment:
what happens if there are two solutions (with equal sums) ?
I assume it does not happen yet but people will be inventive with z3 power of expression.
Also if there are several possible optimization path, z3 will probably be used to choose at some point.
I am not sure what is better but linear trade-off seems weak. Can we minimize something more harder on bigger precision ? square precision ?

Reply:
> what happens if there are two solutions (with equal sums) ?

Can't happen as there is an optimal solution to each circuit (i.e., lowest bit-widths that satisfy the compiler constraints). Let me know if you can generate a case where it happens if you find any!

> I am not sure what is better but linear trade-off seems weak. Can we minimize something more harder on bigger precision ? square precision ?

I didn't understand, could you elaborate further, with an example maybe?